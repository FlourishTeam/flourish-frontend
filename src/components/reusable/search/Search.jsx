import React from 'react';
import PropTypes from 'prop-types';

const Search = ({ onSubmit }) => {
  return (
    <div>
      <label htmlFor="searchInput">Search</label>
      <input 
        id="searchInput"
        type="text" 
        placeholder="search" 
        onSubmit={ onSubmit }
      />
      {/* <button>Submit</button> */}
    </div>
  );
};

Search.propTypes = {
  onSubmit: PropTypes.func.isRequired
};

export default Search;




rf