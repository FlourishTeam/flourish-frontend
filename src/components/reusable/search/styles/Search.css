* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

.Search {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;

    form {
        display: flex;
        justify-content: center;
        align-items: center;
        height: 70px;
        border: 2px inset #ccc;
        padding: 2px;
    }

    .searchInput, .photoMode, .submitButton {
        height: 70px;
        border: none;
        outline-width: 0;
        margin: 5px;
    }

    .searchInput {
        font-size: 2rem;
    }

    .submitButton {
        width: 35px;
        opacity: 0.8;
        background: none;
    }

    .uploaderContainer {
        display: flex;
        flex-wrap: wrap;
        flex-direction: column;
    }

    .photoForm {
     display: flex;
     flex-direction: column;
     height: inherit;
    }

    .fileContainer > img {
        display: none;
        height: auto;
    }

    .uploadThumbnail {
        display: flex;
        justify-content: center;
        align-items: center;
        height: 25vh;
    }
}