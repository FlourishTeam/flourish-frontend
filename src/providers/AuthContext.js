import React, { createContext, useContext, useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import {
  getLogout,
  getVerify,
  postLogin,
  postSignup,
} from '../services/fetches/auth';

const AuthContext = createContext(null);

export const AuthProvider = ({ children }) => {
  const history = useHistory();
  const [session, setSession] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const [signupError, setSignupError] = useState(null);
  const isAuthenticated = !!session;

  useEffect(() => {
    getVerify()
      .then((user) => setSession(user))
      .catch(() => console.log('user not logged in'))
      .finally(() => setLoading(false));
  }, []);
  
  const signup = (name, email, password) => {
    return postSignup(name, email, password)
      .then((user) => setSession(user))
      .catch((err) => setSignupError(err));
  };

  const login = (email, password) => {
    return postLogin(email, password)
      .then((user) => setSession(user))
      .catch((err) => setError(err));
  };

  const logout = () => {
    return getLogout()
      .then(() => {
        setSession(null);
        history.push('/');
      });
  };

  return (
    <AuthContext.Provider
      value={{
        session,
        loading,
        error,
        isAuthenticated,
        signup,
        login,
        logout,
        signupError
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export const useSession = () => {
  const { session } = useContext(AuthContext);
  return session;
};

export const useAuthLoading = () => {
  const { loading } = useContext(AuthContext);
  return loading;
};

export const useAuthError = () => {
  const { error } = useContext(AuthContext);
  return error;
};

export const useSignupError = () => {
  const { signupError } = useContext(AuthContext);
  return signupError;
};

export const useIsAuthenticated = () => {
  const { isAuthenticated } = useContext(AuthContext);
  return isAuthenticated;
};

export const useSignup = () => {
  const { signup } = useContext(AuthContext);
  return signup;
};

export const useLogin = () => {
  const { login } = useContext(AuthContext);
  return login;
};

export const useLogout = () => {
  const { logout } = useContext(AuthContext);
  return logout;
};
