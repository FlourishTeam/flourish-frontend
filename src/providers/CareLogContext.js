import React, { createContext, useContext, useState } from 'react';
import { getMyCareHistoryById } from '../services/queries/getMyCareHistoryById';

export const CareLogContext = createContext(null);

export const CareLogProvider = ({ children }) => {
  const [loading, setLoading] = useState(true);
  const [plantDetails, setPlantDetails] = useState([]);
  const [userPlantId, setUserPlantId] = useState();
  const [careLogItems, setCareLog] = useState([]);
  const [error, setError] = useState(null);

  const renderMyCareHistory = (userId, plantId) => {
    setError(null);
    setLoading(true);

    getMyCareHistoryById(plantId, userId)
      .then(({ data }) => {
        setCareLog(data.getMyCareHistoryById.careLogs);
        setPlantDetails(data.getMyCareHistoryById.plantDetails[0]);
        setUserPlantId(data.getMyCareHistoryById.userPlantId);
        setLoading(false);
      })
      .catch((error) => setError(error.message));
  };

  return (
    <CareLogContext.Provider
      value={{
        loading,
        careLogItems,
        renderMyCareHistory,
        plantDetails,
        userPlantId,
        error,
      }}
    >
      {children}
    </CareLogContext.Provider>
  );
};

export const useUserPlantId = () => {
  const { userPlantId } = useContext(CareLogContext);
  return userPlantId;
};

export const useCareLogLoading = () => {
  const { loading } = useContext(CareLogContext);
  return loading;
};

export const useCareLogItems = () => {
  const { careLogItems } = useContext(CareLogContext);
  return careLogItems;
};

export const usePlantDetails = () => {
  const { plantDetails } = useContext(CareLogContext);
  return plantDetails;
};

export const useCareLogError = () => {
  const { error } = useContext(CareLogContext);
  return error;
};

export const useRenderMyCareHistory = () => {
  const { renderMyCareHistory } = useContext(CareLogContext);
  return renderMyCareHistory;
};
