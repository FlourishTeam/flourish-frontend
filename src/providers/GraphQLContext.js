import fetch from 'cross-fetch';
import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';

const httpLink = createHttpLink({
  uri: 'https://flourish-dev-staging.herokuapp.com/graphql',
  fetch,
});

const client = new ApolloClient({
  link: httpLink,
  cache: new InMemoryCache(),
});

export default client;
