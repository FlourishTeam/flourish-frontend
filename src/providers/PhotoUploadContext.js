import React, { createContext, useContext, useState } from 'react';
import { useHistory } from 'react-router-dom';
import {
  useHandleSearch,
  useSetSearchLoading,
  useSetSearchQuery,
} from './SearchContext';

const API_URL = process.env.REACT_APP_API_URL;
const API_KEY = process.env.REACT_APP_API_KEY;

export const PhotoUploadContext = createContext(null);

export const PhotoUploadProvider = ({ children }) => {
  const [photoMode, setPhotoMode] = useState(false);
  const [pictureUpload, setPictureUpload] = useState(
    'images/placeholder-img.png'
  );
  const [empty, setEmpty] = useState(true);
  const handleSearch = useHandleSearch();
  const setLoading = useSetSearchLoading();
  const setSearchQuery = useSetSearchQuery();

  const history = useHistory();

  const handlePreview = (e) => {
    setPictureUpload(URL.createObjectURL(e.target.files[0]));
    setEmpty(false);
  };

  const handleUpload = (e) => {
    e.preventDefault();
    const files = [...e.target.elements.uploader.files];

    if (files.length > 0) {
      setLoading(true);
      setSearchQuery('');

      const promise = files.map((file) => {
        return new Promise((resolve) => {
          const reader = new FileReader();
          reader.onload = (e) => {
            const res = e.target.result;
            resolve(res);
          };
          reader.readAsDataURL(file);
        });
      });

      Promise.all(promise).then((encodedStr) => {
        const data = {
          api_key: `${API_KEY}`,
          images: encodedStr,
        };

        fetch(`${API_URL}`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(data),
        })
          .then((res) => res.json())
          .then((json) => {
            const plant = json.suggestions[0].plant_name;
            setSearchQuery(plant);
            handleSearch(plant);
          });
      });

      setPhotoMode(false);
      history.push('/search');
    }
  };

  return (
    <PhotoUploadContext.Provider
      value={{
        photoMode,
        setPhotoMode,
        pictureUpload,
        handleUpload,
        handlePreview,
        empty,
      }}
    >
      {children}
    </PhotoUploadContext.Provider>
  );
};

export const usePhotoUploader = () => {
  return useContext(PhotoUploadContext);
};
