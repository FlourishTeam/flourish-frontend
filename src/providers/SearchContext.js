import React, { createContext, useContext, useState } from 'react';
import { plantByName } from '../services/plantByName';

export const SearchContext = createContext(null);

export const SearchProvider = ({ children }) => {
  const [loading, setLoading] = useState(false);
  const [searchResults, setSearchResults] = useState([]);
  const [searchQuery, setSearchQuery] = useState('');
  const [error, setError] = useState(null);

  const handleSearch = (name) => {
    setError(null);
    setLoading(true);

    plantByName(name)
      .then(({ data }) => {
        setSearchResults(data.plantByName);
        setLoading(false);
      })
      .catch((error) => setError(error.message));
  };

  const handleChange = ({ target }) => {
    setSearchQuery(target.value);
  };

  return (
    <SearchContext.Provider
      value={{
        loading,
        searchResults,
        searchQuery,
        handleChange,
        handleSearch,
        setLoading,
        error,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

export const useSearchLoading = () => {
  const { loading } = useContext(SearchContext);
  return loading;
};

export const useSetSearchLoading = () => {
  const { setLoading } = useContext(SearchContext);
  return setLoading;
};

export const useSearchResults = () => {
  const { searchResults } = useContext(SearchContext);
  return searchResults;
};

export const useSearchQuery = () => {
  const { searchQuery } = useContext(SearchContext);
  return searchQuery;
};

export const useHandleChange = () => {
  const { handleChange } = useContext(SearchContext);
  return handleChange;
};

export const useHandleSearch = () => {
  const { handleSearch } = useContext(SearchContext);
  return handleSearch;
};

export const useSearchError = () => {
  const { error } = useContext(SearchContext);
  return error;
};
