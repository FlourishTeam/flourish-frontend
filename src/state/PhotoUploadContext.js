import React, { createContext, useContext, useState } from 'react';
import { useHistory } from 'react-router-dom';
import {
  useHandleSearch,
  useSearchLoading,
  useSetSearchLoading,
} from './SearchContext';

const API_URL = process.env.REACT_APP_API_URL;
const API_KEY = process.env.REACT_APP_API_KEY;

export const PhotoUploadContext = createContext(null);

export const PhotoUploadProvider = ({ children }) => {
  const [photoMode, setPhotoMode] = useState(false);
  const [pictureUpload, setPictureUpload] = useState(
    'images/placeholder-img.png'
  );
  const [empty, setEmpty] = useState(true);
  const handleSearch = useHandleSearch();
  const setLoading = useSetSearchLoading();

  const history = useHistory();

  const handlePreview = (e) => {
    setPictureUpload(URL.createObjectURL(e.target.files[0]));
    setEmpty(false);
  };

  const handleUpload = (e) => {
    e.preventDefault();
    const files = [...e.target.elements.uploader.files];

    if (files.length > 0) {
      setLoading(true);

      const promise = files.map((file) => {
        return new Promise((resolve, reject) => {
          const reader = new FileReader();
          reader.onload = (e) => {
            const res = e.target.result;
            resolve(res);
          };
          reader.readAsDataURL(file);
        });
      });

      Promise.all(promise).then((encodedStr) => {
        const data = {
          api_key: `${API_KEY}`,
          images: encodedStr,
        };

        fetch(`${API_URL}`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(data),
        })
          .then((res) => res.json())
          .then((json) => {
            handleSearch(json.suggestions[0].plant_name);
          });
      });

      setPhotoMode(false);
      history.push('/search');
    }
  };

  return (
    <PhotoUploadContext.Provider
      value={{
        photoMode,
        setPhotoMode,
        pictureUpload,
        handleUpload,
        handlePreview,
        empty,
      }}
    >
      {children}
    </PhotoUploadContext.Provider>
  );
};

export const usePhotoUploader = () => {
  return useContext(PhotoUploadContext);
};
