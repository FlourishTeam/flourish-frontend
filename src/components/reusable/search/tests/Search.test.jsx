import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Search from '../Search';
import { SearchProvider } from '../../../../providers/SearchContext';
import { PhotoUploadProvider } from '../../../../providers/PhotoUploadContext';

describe('Search component', () => {
  afterEach(() => cleanup());

  it('renders Search', () => {
    const { asFragment } = render(
      <SearchProvider>
        <PhotoUploadProvider>
          <Search />
        </PhotoUploadProvider>
      </SearchProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
