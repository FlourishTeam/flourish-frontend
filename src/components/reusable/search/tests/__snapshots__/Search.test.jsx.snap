// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders Search 1`] = `
<DocumentFragment>
  <div>
    <label
      for="searchInput"
    >
      Search
    </label>
    <input
      id="searchInput"
      placeholder="search"
      type="text"
    />
  </div>
</DocumentFragment>
`;
